metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,423 @@
+   *
+   * @param existingNameQuals the `variableName <- qual` tuples which are extracted from the
+   *                          higher levels of the hierarchy of nested `flatMaps`/`maps`,
+   *                          which would make up the enumertions of the for yield expression, later
+   * @param termApply         the whole expression which is to have the name
+   * @param document
+   * @return a tuple of (List of (variableName, Qual), the candidate yield expression)
+   */
+  private def extractForYield(
+      existingNameQuals: List[ForYieldEnumeration],
+      existingConditions: List[ForYieldCondition],
+      termApply: Term.Apply,
+      document: String
+  ): (List[ForYieldEnumeration], List[ForYieldCondition], YieldExpression) = {
+    termApply.fun match {
+      case termSelect: Term.Select
+          if termSelect.name.value == "flatMap" || termSelect.name.value == "map" =>
+        val qual = termSelect.qual
+        val qualString = document.substring(qual.pos.start, qual.pos.end)
+        termApply.args.head match {
+          case termFunction: Term.Function =>
+            extractForYieldOutOfFunction(
+              existingNameQuals,
+              existingConditions,
+              termFunction,
+              qualString,
+              document
+            )
+
+          case termBlock: Term.Block =>
+            termBlock.stats
+              .collectFirst { case termFunction: Term.Function =>
+                extractForYieldOutOfFunction(
+                  existingNameQuals,
+                  existingConditions,
+                  termFunction,
+                  qualString,
+                  document
+                )
+              }
+              .getOrElse(
+                (
+                  existingNameQuals,
+                  existingConditions,
+                  YieldExpression(
+                    document.substring(termApply.pos.start, termApply.pos.end)
+                  )
+                )
+              )
+
+          case _ =>
+            (
+              existingNameQuals,
+              existingConditions,
+              YieldExpression(
+                document.substring(termApply.pos.start, termApply.pos.end)
+              )
+            )
+        }
+      case _ =>
+        (
+          existingNameQuals,
+          existingConditions,
+          YieldExpression(
+            document.substring(termApply.pos.start, termApply.pos.end)
+          )
+        )
+    }
+  }
+
+  def extractForYieldOutOfPotentaialYieldTree(
+      existingNameQuals: List[ForYieldEnumeration],
+      existingConditions: List[ForYieldCondition],
+      yieldTree: Tree,
+      document: String
+  ): (List[ForYieldEnumeration], List[ForYieldCondition], YieldExpression) = {
+    yieldTree match {
+      case termApply: Term.Apply =>
+        extractForYield(
+          existingNameQuals,
+          existingConditions,
+          termApply,
+          document
+        )
+      case termIf: Term.If =>
```

Comment:
if statements should not be translated, there are much more complex and not just syntactical.

We can translate `filter` and `withFilter` function invocations.

Reply:
`if` statements are only handled in cases when the associated `else` is None. I would like to handle the cases when the `else` is an empty collection. That becomes semantically equivalent to `filter` I think. So I guess we would bet able to both keep `if` and `filter`

Reply:
This requires semantic knoledge, we need to know what can be returned if the if statement, the exact type etc. which is not easy to figure out. 

Besides, if else statements are not used often like that and instead most people use filter

Reply:
Well aren’t `Option`s and `Seq`s sort of convertible in this case? because after all `None` is just as much of an empty collection as `List.empty` for example.

Reply:
There is a plenty of other types that you can use for comprehensions with for example `Future` or `Either` and people can define them themselves. Also, people can define their own None even, which will not be the same as the normal one.

```
object None

def hello = {
  list.flatMap{a =>
    if (cond) Some(a)
    else None
  }
}
```
This will not compile, but we will offer the change to for comprehension, which will break everything.

Adding support for `If` gives us a probability that it will work in a couple of cases (which are not used often), while they add a lot of complexity and might break the code.

Reply:
Ok then!